### PR TITLE
Propagate array dimensions from allocations

### PR DESCRIPTION
**Pull request by pchintalapudi**
2D and 3D array dimension lengths cannot change, and thus we can replace loads of array length and array dimension size with their values from when the array was allocated. 1D array dimensions might change, but we can determine if they might using escape analysis. 

~~Currently, the mechanism for identifying array allocations does not work on Windows, due to the insertion of a trampoline function on Windows, which prevents identification by function pointer equivalence from working as intended (this data was collected from #43107). Fixing this for Windows would require overhauling how we handle ccalls during codegen, which is not part of this PR.~~

Identification of arrays is now done through metadata set during codegen, which should theoretically get around the Windows issue.

Depends on #43057 for creating the llvm-alloc-helpers.cpp file

Performance benchmark:

Function:
```
function f()
           s = 0
           a = zeros(Int, 100, 100, 100)
           for i in eachindex(a)
               s += a[i]
           end
           s
       end
```

Master:
Godbolt: https://godbolt.org/z/cj7h3e3Gr
Benchmark:
```
julia> @benchmark f()
BenchmarkTools.Trial: 3179 samples with 1 evaluation.
 Range (min … max):  917.762 μs …  13.191 ms  ┊ GC (min … max):  0.00% … 92.57%
 Time  (median):       1.413 ms               ┊ GC (median):     0.00%
 Time  (mean ± σ):     1.566 ms ± 802.684 μs  ┊ GC (mean ± σ):  10.83% ± 15.99%

  ▅▆█▄        ▂▇▆▃                  ▂▂▃▂  ▃▂    ▁▁  ▂   ▁▃▂   ▂ ▁
  ████▅▅▇█▄▁▁▁████▅▄▆▆▅▁▁▃▄▄▃▁▁▁▁▁▃█████▇███▇▄▄███▇███▅▆███▇▄██ █
  918 μs        Histogram: log(frequency) by time       3.24 ms <

 Memory estimate: 7.63 MiB, allocs estimate: 2.
```

PR:
Godbolt: https://godbolt.org/z/zKWz8dxev
Benchmark:
```
julia> @benchmark f()
BenchmarkTools.Trial: 6350 samples with 1 evaluation.
 Range (min … max):  687.484 μs …   4.911 ms  ┊ GC (min … max): 0.00% …  0.00%
 Time  (median):     702.864 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   782.345 μs ± 188.419 μs  ┊ GC (mean ± σ):  9.65% ± 14.78%

   █
  ▅█▇▃▄▂▃▃▂▂▂▂▂▂▂▂▁▁▂▁▁▁▁▁▁▁▂▂▁▁▁▂▁▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▂▂▃▄▃▃▂▂▂ ▂
  687 μs           Histogram: frequency by time         1.17 ms <

 Memory estimate: 7.63 MiB, allocs estimate: 2.
```

### CODE REVIEW

**Comment by pchintalapudi** [comment]
Since my branches are part of a fork, I've created a PR comparing this one and the allocation hoisting PR immediately before it here: https://github.com/pchintalapudi/julia/pull/4

**Comment by pchintalapudi** [comment]
Now that #43057 has been merged, this PR is ready to be reviewed for rebasing onto the master branch.

**Comment by pchintalapudi** [comment]
Compile time benchmarks:
---
Master:
```
Base  ───────────── 23.796904 seconds
ArgTools  ─────────  4.456852 seconds
Artifacts  ────────  0.102586 seconds
Base64  ───────────  0.105645 seconds
CRC32c  ───────────  0.008005 seconds
FileWatching  ─────  0.108983 seconds
Libdl  ────────────  0.001518 seconds
Logging  ──────────  0.036208 seconds
Mmap  ─────────────  0.085615 seconds
NetworkOptions  ───  0.117357 seconds
SHA  ──────────────  0.181987 seconds
Serialization  ────  0.270087 seconds
Sockets  ──────────  0.321430 seconds
Unicode  ──────────  0.077732 seconds
DelimitedFiles  ───  0.104025 seconds
LinearAlgebra  ────  7.762580 seconds
Markdown  ─────────  0.754207 seconds
Printf  ───────────  0.168311 seconds
Random  ───────────  1.242392 seconds
Tar  ──────────────  0.282602 seconds
Dates  ────────────  1.536529 seconds
Distributed  ──────  0.764613 seconds
Future  ───────────  0.005194 seconds
InteractiveUtils  ─  0.500062 seconds
LibGit2  ──────────  1.366151 seconds
Profile  ──────────  0.257939 seconds
SparseArrays  ─────  2.678179 seconds
UUIDs  ────────────  0.015961 seconds
REPL  ─────────────  3.456523 seconds
SharedArrays  ─────  0.500804 seconds
Statistics  ───────  0.188637 seconds
SuiteSparse  ──────  1.582334 seconds
TOML  ─────────────  0.067595 seconds
Test  ─────────────  0.315075 seconds
LibCURL  ──────────  0.438543 seconds
Downloads  ────────  0.335553 seconds
Pkg  ──────────────  4.502768 seconds
LazyArtifacts  ────  0.001633 seconds
Stdlibs total  ──── 34.712985 seconds
Sysimage built. Summary:
Total ───────  58.511190 seconds
Base: ───────  23.796904 seconds 40.6707%
Stdlibs: ────  34.712985 seconds 59.3271%
    JULIA usr/lib/julia/sys-o.a
Generating REPL precompile statements... 36/36
Executing precompile statements... 1308/1339
Precompilation complete. Summary:
Total ─────── 109.996850 seconds
Generation ──  82.400907 seconds 74.9121%
Execution ───  27.595943 seconds 25.0879%
```
PR:
```
Base  ───────────── 23.928553 seconds
ArgTools  ─────────  4.422442 seconds
Artifacts  ────────  0.103806 seconds
Base64  ───────────  0.106303 seconds
CRC32c  ───────────  0.008324 seconds
FileWatching  ─────  0.110224 seconds
Libdl  ────────────  0.001559 seconds
Logging  ──────────  0.036711 seconds
Mmap  ─────────────  0.086969 seconds
NetworkOptions  ───  0.117836 seconds
SHA  ──────────────  0.184373 seconds
Serialization  ────  0.271228 seconds
Sockets  ──────────  0.322758 seconds
Unicode  ──────────  0.078247 seconds
DelimitedFiles  ───  0.105199 seconds
LinearAlgebra  ────  7.820788 seconds
Markdown  ─────────  0.758388 seconds
Printf  ───────────  0.151605 seconds
Random  ───────────  1.259426 seconds
Tar  ──────────────  0.285620 seconds
Dates  ────────────  1.510449 seconds
Distributed  ──────  0.776262 seconds
Future  ───────────  0.005317 seconds
InteractiveUtils  ─  0.503558 seconds
LibGit2  ──────────  1.377215 seconds
Profile  ──────────  0.258898 seconds
SparseArrays  ─────  2.681444 seconds
UUIDs  ────────────  0.016312 seconds
REPL  ─────────────  3.456168 seconds
SharedArrays  ─────  0.501174 seconds
Statistics  ───────  0.191785 seconds
SuiteSparse  ──────  1.591476 seconds
TOML  ─────────────  0.068235 seconds
Test  ─────────────  0.320033 seconds
LibCURL  ──────────  0.441098 seconds
Downloads  ────────  0.336550 seconds
Pkg  ──────────────  4.571783 seconds
LazyArtifacts  ────  0.001639 seconds
Stdlibs total  ──── 34.851874 seconds
Sysimage built. Summary:
Total ───────  58.781741 seconds
Base: ───────  23.928553 seconds 40.7075%
Stdlibs: ────  34.851874 seconds 59.2903%
    JULIA usr/lib/julia/sys-o.a
Generating REPL precompile statements... 36/36
Executing precompile statements... 1308/1339
Precompilation complete. Summary:
Total ─────── 109.750784 seconds
Generation ──  81.986599 seconds 74.7025%
Execution ───  27.764185 seconds 25.2975%
```

**Comment by pchintalapudi** [comment]
# Other notable examples of optimization improvement:
---
## Example 2: doc/src/manual/performance-tips.md:952-963 loopinc

### Master:
Godbolt: https://godbolt.org/z/oT9Wa4784

### PR:
Godbolt: https://godbolt.org/z/r1or75rzK

## Example 3: doc/src/manual/performance-tips.md:969-984 loopinc_prealloc

### Master:
Godbolt: https://godbolt.org/z/7Ed4xd3o9

### PR:
Godbolt: https://godbolt.org/z/G6adnc69f

---
Both examples show complete removal of the bounds check due to length propagation, and LLVM then recognizes that all of the loads from the array can be forwarded and eliminates them all. When combined with #43548 and its ability to remove arrays that are never loaded from, and if xinc(!) is not marked noinline, LLVM will either vectorize loopinc (resulting in a massive speedup) or outright precompute the sum in loopinc_prealloc (another O(n) -> O(1) transformation)

**Comment by vchuravy** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/43487#issuecomment-1008126016) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/1f3cb02_vs_acb6e16/report.html).

**Comment by vchuravy** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

File: src/ccall.cpp [hunk 1]
```diff
@@ -1779,6 +1779,33 @@ static jl_cgval_t emit_ccall(jl_codectx_t &ctx, jl_value_t **args, size_t nargs)
...
+    } else if (is_libjulia_func(jl_alloc_array_1d) || is_libjulia_func(jl_alloc_array_2d)
+            || is_libjulia_func(jl_alloc_array_3d)) {
+        jl_cgval_t retval = sig.emit_a_ccall(
+                ctx,
+                symarg,
+                argv,
+                gc_uses,
+                static_rt);
+        if (auto array_alloc = dyn_cast<CallInst>(retval.V)) {
+            array_alloc->addAttribute(AttributeList::ReturnIndex, Attribute::NoAlias);
```
Comment (vchuravy) [inline src/ccall.cpp:1791]: Also `NonNull` to mimick `julia.gc_alloc_obj`?

Comment (pchintalapudi) [reply src/ccall.cpp:1791]: It's already marked `NonNull` without explicitly adding the attribute here.

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/43487#issuecomment-1008458298) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/c1c20a8_vs_f7bb391/report.html).

**Comment by vchuravy** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/43487#issuecomment-1009182101) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/12b9a1b_vs_7b62896/report.html).

**Review by vchuravy — APPROVED** [review]

**Comment by vchuravy** [comment]
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/43487#issuecomment-1009530271) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/b92fd65_vs_a7e4f6d/report.md).

**Comment by vchuravy** [comment]
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

**Comment by vchuravy** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by vchuravy** [comment]
If nanosoldier comes back clean I am going to merge this.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/43487#issuecomment-1012679734) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/19b2f65_vs_294b0df/report.md).

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/43487#issuecomment-1012679938) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/cc79187_vs_294b0df/report.html).

**Comment by vtjnash** [comment]
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

**Comment by vtjnash** [comment]
I am not certain if we want to move forward with tricks like this. The goal for Array was to replace it with something simpler implemented in Julia, which would hopefully allow the compiler to "see" this, without needing to make it a special case again in codegen (we had some code like this many years ago).

**Comment by vchuravy** [comment]
>  The goal for Array was to replace it with something simpler implemented in Julia, which would hopefully allow the compiler to "see" this

I think both Prem and I agree with that notion. But it is also the case that the move from `Array` to `Buffer` hasn't happened yet (and we are adding things like ImmutableArrays). So it might be a case of the perfect is the enemy of the good.

**Comment by pchintalapudi** [comment]
I think there's another point to consider, which is that if Array is implemented in Julia as something like:
```
struct Array
    ptr::Buffer
    dims::Tuple
end
```
we cannot stack-allocate the internal buffer of the array even with interprocedural escape analysis. The reasoning behind that is linked to the possible reshape method below:
```
function reshape(a::Array, dims::Tuple)
    reshaped = Array()
    reshaped.dims = copy(dims)
    reshaped.ptr = a.ptr
    reshaped
end
```
Here, the lifetime of `a.ptr` outlives the lifetime of a itself. If reshape is called from another method and not inlined, then the other method could not stack allocate `a.ptr` unless it knew that `a.ptr` wasn't captured in a way like that. Right now this is not an issue because whenever we extend the lifetime of the internal data pointer beyond the outer, we also perform a ccall and so the outer object is also escaped. If we don't need a ccall to extend that pointer's lifetime, to fully stack allocate arrays we would need to know if a method was capturing the outer object _as well as_ if the method was capturing a pointer inside the outer object, which could become infeasible with larger objects or larger numbers of arguments. The only way I see around this would be to make the Array type completely opaque to Julia code; that is, no Julia code can see or modify the internal `ptr` field. Unfortunately, I suspect this brings us back to Array being special-cased again, because with an opaque type the Julia compiler couldn't see into the type to make constant propagation happen.

**Comment by pchintalapudi** [comment]
I think there's another point to consider, which is that if Array is implemented in Julia as something like:
```
struct Array
    ptr::Buffer
    dims::Tuple
end
```
we cannot stack-allocate the internal buffer of the array even with interprocedural escape analysis. The reasoning behind that is linked to the possible reshape method below:
```
function reshape(a::Array, dims::Tuple)
    reshaped = Array()
    reshaped.dims = copy(dims)
    reshaped.ptr = a.ptr
    reshaped
end
```
Here, the lifetime of `a.ptr` outlives the lifetime of a itself. If reshape is called from another method and not inlined, then the other method could not stack allocate `a.ptr` unless it knew that `a.ptr` wasn't captured in a way like that. Right now this is not an issue because whenever we extend the lifetime of the internal data pointer beyond the outer, we also perform a ccall and so the outer object is also escaped. If we don't need a ccall to extend that pointer's lifetime, to fully stack allocate arrays we would need to know if a method was capturing the outer object _as well as_ if the method was capturing a pointer inside the outer object, which could become infeasible with larger objects or larger numbers of arguments. The only way I see around this would be to make the Array type completely opaque to Julia code; that is, no Julia code can see or modify the internal `ptr` field. Unfortunately, I suspect this brings us back to Array being special-cased again, because with an opaque type the Julia compiler couldn't see into the type to make constant propagation happen.

**Comment by vtjnash** [comment]
That is not really a new point, but is already a problem: just replace "struct Array" with any existing type, such as SubArray, Ref, Tuple, etc., and you have provided a rough definition for escape analysis already.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/43487#issuecomment-1013261451) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/f912f69_vs_b57d2e1/report.md).

**Comment by pchintalapudi** [comment]
I think the crux of stack allocating the buffer data relies on the lifetime of the data pointer always being <= the lifetime of the containing object. If we had a way of communicating that sort of relationship within the language to the compiler, we could extend stack allocation to subfields of generic objects as well. That seems fairly risky though, as there isn't really a clear way to warn someone when they're about to violate that assumption, and it's really really easy to violate that assumption.

Of course, this is also a pretty niche case that relies on a lot of different conditions and analyses being present (we don't have either stack allocation for arrays yet or IPO escape analysis, although I believe both are coming soon), so it's hard to say whether such an optimization is even worth the effort yet. That being said, as Valentin says this might be more worth it as a optimization now while we don't have a full-Julia implementation of Array, especially since basic array stack allocation is 4-5 PRs down the line from this one (#43573) and depends on the array identification machinery proposed here. In the future, we might end up purging the arraysize/arraylen codegen functions regardless if it's entirely specified in Julia, so we can drop the codegen modifications as they get deleted.

**Comment by PallHaraldsson** [comment]
This has been approved by 1/4 reviewers. Is it ok to merge? Are the "3 failing checks" false alarms? This seems like important work blocking very important stack-allocation work PR.

**Pull request closed without merging**